[SPIRV] GPU intrinsics

JonChesterfield opened the pull request:
Introduce __builtin_gpu builtins to clang and corresponding llvm.gpu intrinsics in llvm for abstracting over minor differences between GPU architectures, and use those to implement a gpuintrin.h instantiation to support compiling llvm libc to spirv64--.

Motivated by discussion at https://discourse.llvm.org/t/rfc-spirv-ir-as-a-vendor-agnostic-gpu-representation/85115 and RFC for this specifically at https://discourse.llvm.org/t/rfc-llvm-gpu-builtins-for-target-agnostic-code-representation/85181.

These are not named llvm.spirv because there are no spirv specific semantics involved. They're deliberately the same small abstraction over targets used by llvm libc already. Essentially this patch allows us to postpone choosing a target architecture for libc until JIT time. 

There is some refactoring to be done if this lands - moving some work out of CGBuiltin.cpp, simplifying gpuintrin.h, adjusting openmp's codegen to use these instead of devicertl magic functions.

JonChesterfield committed:
[SPIRV] GPU intrinsics

jhuber6 commented on clang/include/clang/Basic/Builtins.td:
```diff
@@ -4770,6 +4770,35 @@ def GetDeviceSideMangledName : LangBuiltin<"CUDA_LANG"> {
 }
 
+// GPU intrinsics
+class GPUBuiltin<string prototype> : Builtin {
```
We need a way to make these only emitted on 'gpu' compilations.

JonChesterfield replied:
Not particularly. Intrinsics aren't required to lower to non-error states on all targets anyway. If someone wants to use these to write x64 code with implied simt semantics and deal with that later, cool - e.g. ipsc might like it.

jhuber6 commented on clang/lib/Headers/spirvintrin.h:
```diff
@@ -0,0 +1,182 @@
+//===-- spirvintrin.h - SPIRV intrinsic functions ------------------------===//
```
The header stuff is separate from introducing these builtins, they should be tested in isolation.

JonChesterfield replied:
That would look a lot like https://github.com/llvm/llvm-project/pull/131164. The gpuintrin.h header is checked by the existing test scaffolding with spirv added, I added more for the builtins/intrinsics in this patch.

jhuber6 commented on llvm/lib/Transforms/Scalar/LowerGPUIntrinsic.cpp:
```diff
@@ -0,0 +1,501 @@
+
+  // Note: "__oclc_ABI_version" is supposed to be emitted and initialized by
+  //       clang during compilation of user code.
+  StringRef Name = "__oclc_ABI_version";
```
I hate this thing.

JonChesterfield replied:
That's the lowering of the intrinsic to amdgpu, it's gonna be amdgpu specific. Prior to the lowering which can be at JIT time you've got llvm.gpu.num_threads_x or __builtin_gpu_num_threads_x.

arsenm replied:
Really should be part of the target OS version in the triple.

But checking this here is a sign this isn't fundamentally addressing the issue

jhuber6 reviewed (commented):
So this is basically a serialization of the headers in `gpuintrin.h` as IR? I'll need to defer to the people more familiar with the backend / IR layer but I have a few comments.

JonChesterfield commented:
The intrinsic set matching the gpuintrin.h set is important context for why I chose these ones. That's why the header is included here. The patch would read better with the spirv stub in place first but it wouldn't change the semantics.

The idea is to translate the core gpuintrin functions into IR, then later expand them out to what would have been done in the front end if the target was assumed up front. A side effect is the dispatch to different architectures is done by an IR pass instead of the preprocessor so the gpuintrin.h headers could become significantly shorter. Other side effects are we have some target independent intrinsics that we could use elsewhere in the compiler.

AlexVlx commented:
I'm not convinced we should do this, as I have a bunch of concerns:

- it's intrusive and duplicates work already done by [libclc](https://github.com/llvm/llvm-project/tree/main/libclc);
- [SPIRV already has intrinsics](https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/IR/IntrinsicsSPIRV.td)
- it squats on some prime naming real estate
- it is extremely chunky
- not everything should end up in Clang, even though it is individually "convenient"

I am equally uneasy with the cavalier attitude towards differences between various GPU and GPU-like targets.

jhuber6 commented:
> not everything should end up in Clang, even though it is individually "convenient"

We should at least have some builtins for those SPIR-V intrinsics you listed. I think right now we only have `reflect`.

AlexVlx commented:
> > not everything should end up in Clang, even though it is individually "convenient"
> 
> We should at least have some builtins for those SPIR-V intrinsics you listed. I think right now we only have `reflect`.

Probably, I think there's a sense that in general it's desirable to mirror all target intrinsics to Clang builtins by default, except extremely exotic ones.

jhuber6 commented:
> > > not everything should end up in Clang, even though it is individually "convenient"
> > 
> > 
> > We should at least have some builtins for those SPIR-V intrinsics you listed. I think right now we only have `reflect`.
> 
> Probably, I think there's a sense that in general it's desirable to mirror all target intrinsics to Clang builtins by default, except extremely exotic ones.

At a high level, if those work, we could just use them in `spirvintrin.h`. Do you happen to know what the backend does with them? They seem to be like what's going on here but I don't know when the target specific magic gets baked in.

JonChesterfield commented:
If the name llvm.gpu is a stumbling block, how about llvm.offload? Will add JD to the reviewers

arsenm commented on llvm/include/llvm/IR/IntrinsicsAMDGPU.td:
```diff
@@ -150,6 +150,8 @@ defm int_amdgcn_workitem_id : AMDGPUReadPreloadRegisterIntrinsic_xyz;
 defm int_amdgcn_workgroup_id : AMDGPUReadPreloadRegisterIntrinsic_xyz_named
                                <"__builtin_amdgcn_workgroup_id">;
 
+defm int_amdgcn_grid_size : AMDGPUReadPreloadRegisterIntrinsic_xyz;
```
This is dead code

JonChesterfield replied:
It's not dead, but that's unimportant - the patch could be reworked to not include it and I'll do that in the spirit of increasing consensus

arsenm commented on llvm/include/llvm/IR/Intrinsics.td:
```diff
@@ -2861,6 +2861,69 @@ def int_experimental_convergence_anchor
+//===------- GPU Intrinsics -----------------------------------------------===//
+
+class GPUIntrinsic<LLVMType ret_type, string name>
+  : DefaultAttrsIntrinsic<[ret_type],
+                          [],
+                          [NoUndef<RetIndex>, IntrNoMem, IntrSpeculatable]>,
+                          ClangBuiltin<name>;
```
As I mentioned on the thread I do not think this should be the first solution to reach for.

For example as-is this immediately breaks running AMDGPUAttributor, which will now incorrectly not detect uses of the underlying intrinsics hidden behind these wrappers. These wrappers need to have the same semantics as an unknown external call (as in, this should not use DefaultAddrsIntrinsic)

And need a test showing that AMDGPUAttributor infers none of the implicit argument attributes if there is a call to any of these

JonChesterfield replied:
It doesn't break amdgpu attributor. The attributor will correctly completely ignore these intrinsics, as it should do, unless at some point we decide the amdgpu attributor should start working in terms of target agnostic intrinsics at which point it'll work fine. We don't need an extra test showing that the amdgpu attributor works correctly when shown unknown intrinsics, because if it doesn't have that property already it's already severely broken.

arsenm replied:
https://godbolt.org/z/9Gv4jsW3n even the undecorated fake.use infers everything

arsenm commented on llvm/lib/Transforms/Scalar/LowerGPUIntrinsic.cpp:
```diff
@@ -0,0 +1,501 @@
+  //       clang during compilation of user code.
+  StringRef Name = "__oclc_ABI_version";
+  auto *ABIVersionC = M.getNamedGlobal(Name);
+  if (!ABIVersionC) {
+    // In CGBuiltin, we'd have to create an extern variable to emit the load for
+    // Here, we can leave the intrinsic in place and it'll get lowered later
+    return nullptr;
+  }
```
This should not be mandatory

JonChesterfield replied:
It isn't. That's what the return branch means. No magic variable saying where the value is to be found, leave the intrinsic in place until that variable shows up later. If it never shows up, well, we haven't chosen a code object format and can't codegen.

arsenm commented on llvm/lib/Transforms/Scalar/LowerGPUIntrinsic.cpp:
```diff
@@ -0,0 +1,501 @@
+
+    for (auto *U : make_early_inc_range(Intr->users())) {
+      if (auto *CI = dyn_cast<CallBase>(U)) {
+        if (CI->getCalledFunction() == Intr)
```
This can be an assert

JonChesterfield replied:
I have issues with `__oclc_ABI_version`. However I'm not trying to fix that inline here, only to provide the same lowering we use already. I do like the new feature that IR without that present would be meaningful without adding an extern reference into it (because the load would remain hidden in this intrinsic, which is why this doesn't inject an extern reference in CGBuiltin style).

arsenm commented on llvm/lib/Transforms/Scalar/LowerGPUIntrinsic.cpp:
```diff
@@ -0,0 +1,501 @@
+      continue;
+
+    for (auto *U : make_early_inc_range(Intr->users())) {
+      if (auto *CI = dyn_cast<CallBase>(U)) {
```
```suggestion
      if (auto *CI = dyn_cast<CallInst>(U)) {
```

arsenm commented on llvm/lib/Transforms/Scalar/LowerGPUIntrinsic.cpp:
```diff
@@ -0,0 +1,501 @@
+}
+
+// Replace intrinsic call with a linear sequence of instructions
+typedef Value *(*builder)(Module &M, IRBuilder<> &Builder, Intrinsic::ID from,
+                          CallBase *CI);
```
using function_ref

arsenm commented on llvm/lib/Transforms/Scalar/LowerGPUIntrinsic.cpp:
```diff
@@ -0,0 +1,501 @@
+
+  Triple TT(M.getTargetTriple());
+
+  if (!TT.isAMDGPU() && !TT.isNVPTX()) {
+    return Changed;
+  }
```
The usual convention for ad-hoc IR lowering passes like this is to have some hook in TargetLowering to process the intrinsic. I'm not sure where you're planning on running this expansion, or if it's permissible to pull in codegen there

JonChesterfield replied:
The backend should run it so that the JIT model expands it to meaningful code. However if the only place run is the backend, all the programming languages will refuse to use it for amdgpu or nvptx targets. So I want to _also_ run it early in the clang pipeline where targetting a specific language, and have the backend run handle it for whatever spirv64-- was converted into and for the O0 or llvm passes disabled case.

arsenm commented on llvm/test/CodeGen/SPIRV/gpu_intrinsics.ll:
```diff
@@ -0,0 +1,427 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --check-attributes
+; RUN: opt -S -mtriple=amdgcn-- -passes=lower-gpu-intrinsic < %s | FileCheck %s --check-prefix=AMDGCN
```
Should not be testing amdgpu handling in test/CodeGen/SPIRV. Duplicate the test.

Also this is just an IR->IR test, should go in test/Transforms/LowerGPUIntrinsics

JonChesterfield replied:
I don't love copy&paste the test since the input IR is identical between the two and they'll definitely drift out of sync but sure, I'll copy&paste the thing as requested. Dumping them in a new directory to slightly improve the chances that people editing one remember to edit the other.

arsenm commented on llvm/test/CodeGen/SPIRV/gpu_intrinsics.ll:
```diff
@@ -0,0 +1,427 @@
+
+define void @thread_suspend() {
+; AMDGCN-LABEL: @thread_suspend(
+; AMDGCN-NEXT:    call void @llvm.amdgcn.s.sleep(i32 2)
```
why 2

arsenm commented on llvm/lib/Transforms/Scalar/LowerGPUIntrinsic.cpp:
```diff
@@ -0,0 +1,501 @@
+  lowerFunction NVPTX;
+};
+
+using namespace Intrinsic;
```
```suggestion
```
The unqualified intrinsic names below are a bit confusing

arsenm reviewed (commented):
As I menti

michalpaszkowski commented on llvm/test/CodeGen/SPIRV/gpu_intrinsics.ll:
I need to read through the RFC and the changes in the PR yet but I think `llvm/test/CodeGen/SPIRV/` is not really the best place for this test as it is not testing the SPIR-V backend.

michalpaszkowski commented on llvm/lib/Transforms/Scalar/LowerGPUIntrinsic.cpp:
```diff
@@ -0,0 +1,501 @@
+//
+// Lower the llvm.gpu intrinsics to target specific code sequences.
+// Can be called from clang if building for a specific GPU or from the backend
+// as part of a SPIRV lowering pipeline. Initial pass can lower to amdgcn or
```
I don't understand how this pass fits into the SPIR-V lowering pipeline. Is this confusing [SPIR and SPIR-V](https://en.wikipedia.org/wiki/Standard_Portable_Intermediate_Representation)? SPIR being the format based on LLVM IR (although on an old version of LLVM 3.4)? SPIR-V being a custom [Khronos-defined format](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html).

[While SPIR-V easily maps to LLVM IR](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/main/docs/SPIRVRepresentationInLLVM.rst), it is far from being a standardized mapping and the actual translation is usually handled by either [SPIRV-LLVM-Translator](https://github.com/KhronosGroup/SPIRV-LLVM-Translator) or the in-tree SPIR-V backend.

Could you please comment on how this pass and the changes would fit into the current SPIR-V code generation flow through the SPIR-V backend?

JonChesterfield replied:
This is all written in terms of the spirv64-unknown-unknown triple. The in tree backend writes that out as SPV which sort of punts the problem onward to someone else, or the spirv-translator converts the spirv--unknown-unknown LLVM IR into SPV. Then the spirv-translator or something else turns the SPV into something executable.

The spirv backend should stash these intrinsics in the SPV file like any other ones. It might need a patch to do so or we might get default handling for llvm prefixed intrinsics. That's the point really - we pass information about the SIMT computation through the SPV onward to whatever is dealing with that later.

JonChesterfield replied:
As in the commit message, I want spirv to be usable as a target agnostic representation for gpu compute. https://discourse.llvm.org/t/rfc-spirv-ir-as-a-vendor-agnostic-gpu-representation/85115. Pretty much all it needs for that is a way to talk about simt intrinsics and a lowering to the vendor.

I think it is _deeply_ silly to have lots of intrinsics named after languages which mean the same thing and want to fold them together into the llvm core. So far I'm not getting a lot of love for that but I concede that sweeping cleanup isn't strictly necessary, being technically much cleaner doesn't always means politically tractable.

The builtins and intrinsics introduced here are target agnostic. I think people would be much happier if I'd named them llvm.spirv or llvm.openmp but given they have nothing to do with spirv or openmp, aside from being useful for both, I'm arguing against that.

Spir64-unknown-unknown doesn't say anything about amdgpu or nvptx or intel. Nor should it. It really doesn't say much about the GPU architecture. That's why it's viable as a target to JIT into amdgpu or nvptx later. That's the goal of spirv64, right? Single shader IR that runs on arbitrary GPUs?

I don't particularly care how these intrinsics are represented in SPV, provided they can be raised back onto these llvm intrinsics in the translator. The name llvm is appropriate as that's where the semantics for these intrinsics is defined. I don't know what distinction you're drawing between abstract function calls and intrinsics, terminology is not consistent across compilers here.

MrSidims replied:
First of all I agree, that having a common interface between various frontends is a good idea. Yet as from what I see now - the intrinsics added in the PR are not common, but tuned only for AMDGPU and NVPTX. And just like @michalpaszkowski I don't understand, what place SPIR-V has for those intrinsics. 

>The spirv backend should stash these intrinsics in the SPV file like any other ones. It might need a patch to do so or we might get default handling for llvm prefixed intrinsics. That's the point really - we pass information about the SIMT computation through the SPV onward to whatever is dealing with that later.
>I don't know what distinction you're drawing between abstract function calls and intrinsics, terminology is not consistent across compilers here.

Please correct me if I understood this part incorrectly. Is your proposal to store `llvm.gpu` intrinsic call in SPIR-V module as is if there is no appropriate SPIR-V instruction for it?

JonChesterfield replied:
They're optionally lowered to amdgpu or to nvptx as they're the only two architectures in tree, or they come out the other end in IR to be handled later in the pipeline. I'd like lowering for intel too but that will need to be out of tree until intel comes in tree. The semantics are hopefully fairly general purpose but I expect they'll need iterative improvement as one learns more about other targets - the first draft is the basis set of llvm libc.

I don't really understand the spirv pipeline. The one I've been working with is the amdgpu / hip one, which passes llvm intrinsics through unchanged, and ends up creating amdgcn-- LLVM IR to feed back into the backend. My hope is that in the first instance the SPIR-V module will contain the llvm.gpu intrinsic call, and then sometime later it'll pick up a proper SPIRV name of some sort, but where conversion back to llvm with that intrinsic in it works losslessly.

michalpaszkowski replied:
The problem is that SPIR-V in itself is an architecture (although a synthetic one) and the SPIR-V backend is what lowers LLVM IR to this architecture. There are proprietary SPIR-V consumers which are not LLVM-based. For this reason it might be unreasonable to expect those to accept `llvm.` calls and it would be good to plan early how the new `llvm.gpu` intrinsics would be handled in the SPIR-V backend case by case.

MrSidims replied:
> I'd like lowering for intel too but that will need to be out of tree until intel comes in tree.

To clarify: while me and Michal are working for Intel, our concerns are not about lowering to some Intel's internal stuff, but about lowering to SPIR-V, which is supported as intermediate representation for Vulkan, OpenCL etc by multiple vendors.  I can't speak for others, but (judging by amount of PRs I've been looking in LLVM's SPIR-V backend) apart of Intel at least Google has interest in developing SPIR-V backend. So let me actually invite @sudonatalie and @Keenuts to the discussion.

> which passes llvm intrinsics through unchanged, and ends up creating amdgcn-- LLVM IR to feed back into the backend

I assume it only works, when the frontend and the backend are built on top of the same LLVM version, right? Otherwise you would face an issues like: a. intrinsics semantics can change depending on LLVM version; b. rules of intrinsics' name can change from version to version (for example LLVM's switch from typed to untyped pointers has affected the names).

So talking about SPIR-V toolchains, as it's hard to know ahead of time, who will consume the generated SPIR-V - we must not have compiler generated external symbols (like intrinsics) in the module as the backend might not be able to resolve them.

JonChesterfield commented:
This patch probably does too many things and it's implemented in a non-conventional fashion. I think that does a decent job of having something to upset most people that have looked at it. I note that reviewers are primarily complaining about disjoint aspects of the patch so a more optimistic reading could be that most of you love some of it and hate part of it, and that the links to the RFCs may have been too subtle.


The functionality I need here, independent of the implementation, is:

1. Represent GPU SIMT constructs in LLVM IR without up front specifying a target ISA

What I _want_ beyond that is:

1. Be able to lower that to target specific just out of clang, so they're free to use on a specific target
2. Easy, clean ways to test the lowering (i.e. please let me do this in IR)

If I can't have cheap lowering the openmp crowd will insist on amdgpu and nvptx specific paths that duplicate the lowering that llvm already knows how to do, and so will the hip crowd, because expanding the intrinsics in the backend will induce overhead. Which will primarily stop me removing the target specific dispatch from the language implementations. Not on the critical path, just stops me simplifying things.

This patch is conceptually a few things.

1. My first step towards making spirv64-unknown-unknown usable across vendors
4. Target independent intrinsics for the read_first_lane, ballot etc primitives of the SIMT model
5. Intrinsics for the various grid model accessors
6. A few misc things (exit, suspend) which are kind of niche
7. Doing target specific lowering in an IR pass
8. Doing lowering for _two different_ targets in the _same_ IR pass


I want the first point. Compile code once then JIT it onto some arbitrary GPU architecture? Love it. Noone else seems keen but that's fine, I think some end users will like that functionality. It seems to have been some of the motivation behind spirv too.

Intrinsics for ballot are necessary in the same way that convergent needs to be represented in the IR. Presently the different targets have their own intrinsics for these, because they really should be intrinsics. That I think is absolutely uncontentious. It's absurd that we have multiple different intrinsics that mean the same thing for a common concept that we want to manipulate in IR.

The grid builtins are convenient for the opencl/cuda programming model. They don't matter much to IR manipulation. I want them here because the cost is tiny, most are directly mapped to a register read and the others to some trivial code sequence, and it frees one from the endless squabble over which special compiler-rt is the bestest.

The exit/suspend stuff is niche. Maybe I can't get those through review. They're here because I've gone with "what would it take to run libc on Intel GPUs, assuming Intel won't put target specific code in upstream". I can make them both nops under spirv for the time being.

I've done the lowering in IR because that's the right and proper thing to do, fully aware that most people disagree with me on that. We get trivial testing, composability, not writing the thing twice for SDag and GlobalISel, and the ability to lower right out of clang where the target is known early. So OpenMP that uses these intrinsics and targets nvptx64-- directly pays zero cost. Likewise amdgpu-- as the target will immediately turn into the target intrinsics and interact with the amdgpu attributor as if this was never here. The pass should be scheduled early after clang as an optimisation and run unconditionally in the backend. Some people might recognise this strategy from my variadic lowering pass.

I've lowered amdgpu and nvptx in the same IR pass, which is heresy, because I want to emphasise that these are not semantically clever or novel things, that can be rewritten to target specific things whenever is convenient. That seems to have been misread above. I also want it to be very obvious to Intel exactly where they need to add a column of function pointers to have this work on their backend. An extra target gets a column, an extra intrinsic gets a row, fill in all the grid spaces. I like dumb code that walks over tables.


However. This has not sailed cleanly through review, and I'm totally willing to compromise the pretty implementation to get the functionality of build once, JIT to any GPU. If I have to copy&paste files across directories and delete parts so be it. Likewise I'll cheerfully leave a load of conceptually redundant target dispatch cruft in libc and openmp if you guys won't let me lower these in an IR pass shortly after clang.

JonChesterfield commented:
I've dropped the clang headers part from this patch, rewritten to elide the grid intrinsic and moved the test. Matt suggested llvm.simt as the prefix which I like very much more than llvm.gpu but the rename is probably going to take a moment.

Turns out I do need to update the amdgpu attributor pass to know about the new names which is upsetting but fine.

JonChesterfield commented:
Patch opposed internally. Will see if I can think of an alternative.
